mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp

```diff
@@ -160,6 +160,23 @@ class CmpFOpConversion : public OpConversionPattern<arith::CmpFOp> {
   }
 };
 
+/// Add an explicit C-style cast operation to unsigned, to force
+/// the unsigned interpretation of a signless integer.
+Value castSignlessToUnsigned(ConversionPatternRewriter &rewriter, Location loc,
+                             Value operand) {
+  auto intType = dyn_cast<IntegerType>(operand.getType());
+  if (!intType)
+    return operand;
+
+  IntegerType::SignednessSemantics operandSignedness = intType.getSignedness();
+  if (operandSignedness == IntegerType::SignednessSemantics::Signless) {
```

Comment:
I think arith only allows `Signless`, so maybe we can have an assertion here. If this would be changed, the function might need adaption to cast si -> ui, and we would notice from the assertion.

Reply:
I think the type here is afte type conversion, and there were discussions on EuroLLVM to make emitc have signed instead of signless. So I would interpret both signless and signed here for now (similar as it is done for the arith.addi conversion)

Reply:
Correct, types are after conversion. Just to make sure, the operands found in the adaptor already have converted types, there's no need to convert them more, right?

Ideally I'd like this function to be the only point that relies on the signless interpretation as signed, so I'll make it a tiny bit more generic so it's called for all operands of all operations, and it inserts casts where appropriate.

Reply:
Yes, the adaptor contains the converted values. This is the reason for the adaptor: It's a container of converted Values that expose the same interface as the original ops, so you can still refer to operands by name.

Reply:
After discussing that with @TinaAMD, I came to think that I'll just go for a signedness mismatch test within the `uitofp` conversion instead of a separate function. The upstream arith.addi you mentioned already has a good implementation of this test, so we'll be able to factor it into a function once we bump this fork.